sigstore/oidc.py

```diff
@@ -57,6 +58,107 @@ class _OpenIDConfiguration(BaseModel):
+
+        self._raw_token = raw_token
+
+        # NOTE: The lack of verification here is intentional, and is part of
+        # Sigstore's verification model: clients like sigstore-python are
+        # responsible only for forwarding the OIDC identity to Fulcio for
+        # certificate binding and issuance.
+        try:
+            self._unverified_claims = jwt.decode(
+                self._raw_token, options={"verify_signature": False}
+            )
+        except jwt.InvalidTokenError as exc:
+            raise IdentityError("invalid identity token") from exc
+
+        self._issuer: str = self._unverified_claims.get("iss")  # type: ignore
+        if self._issuer is None:
+            raise IdentityError("Identity token missing the required `iss` claim")
+
+        aud = self._unverified_claims.get("aud")
+        if aud is None:
+            raise IdentityError("Identity token missing the required `aud` claim")
+        if aud != _DEFAULT_AUDIENCE:
+            raise IdentityError(
+                f"Audience should be {_DEFAULT_AUDIENCE!r}, not {aud!r}"
+            )
+
+        # When verifying the private key possession proof, Fulcio uses
+        # different claims depending on the token's issuer.
+        # We currently special-case a handful of these, and fall back
+        # on signing the "sub" claim otherwise.
+        identity_claim = _KNOWN_OIDC_ISSUERS.get(self.issuer)
+        if identity_claim is not None:
+            if identity_claim not in self._unverified_claims:
+                raise IdentityError(
+                    f"Identity token missing the required {identity_claim!r} claim"
+                )
+
+            self._identity = str(self._unverified_claims.get(identity_claim))
+        else:
+            try:
+                self._identity = str(self._unverified_claims["sub"])
+            except KeyError:
+                raise IdentityError("Identity token missing the required 'sub' claim")
+
+    @property
+    def identity(self) -> str:
+        """
+        Returns this `IdentityToken`'s underlying subject.
+        """
+        return self._identity
+
+    @property
+    def issuer(self) -> str:
+        """
+        Returns a URL identifying this `IdentityToken`'s issuer.
+        """
+        return self._issuer
+
+    @property
+    def federated_issuer(self) -> str | None:
+        """
+        Returns a URL identifying the "federated" issuer for this `IdentityToken`.
+
+        This is nominally an **implementation detail** of a Sigstore deployment,
+        but is exposed because some applications may find it useful to filter/exclude
+        identity tokens by their "original" issuer, even if the token has been
+        forwarded through a federated provider (e.g., a Dex instance).
+        """
+        federated_claims = self._unverified_claims.get("federated_claims")
+        if federated_claims is None:
+            return None
+
+        return str(federated_claims.get("connector_id"))
```

Comment:
An unlikely occurrence I suppose but if the key isn't found I think this returns the string `"None"`.

Reply:
Yeah, that's bad. Will fix!

Reply:
Refactored and made more sound with eb14330..e03db4f.